Clarify how SUBSCRIBE filters work, allow Absolute start before Latest (#927)
moq-wg/moq-transport: draft-ietf-moq-transport.md

The subscriber specifies a filter on the subscription to allow
the publisher to identify which objects need to be delivered.

> All filters have a Start Location and an optional End Group.  Only objects

Reviewer: wonder why is it EndGroup and not End Location?

Author: #593 removed End Object from Absolute Range.

Reviewer: Hmm . I missed this somehow. Thanks for sharing the PR. That PR does say this "Specifies a closed subscription starting at StartObject
in StartGroup and ending at EndObject in EndGroup."

I wonder if removing the EndObject was resolving any issue. Sorry for missing it earlier.

Author: @ianswett do you have the linked issue for that?

published or received via a subscription having Locations greater than or

> equal to Start and strictly less than or equal to the End Group (when

Reviewer: ```suggestion
equal to Start (unless Start Location is {0,0} ) and strictly less than or equal to the End Group, (when
```

Author: I don't understand why {0,0} would be an exception?

Reviewer: since it represents AllObjects and that would allow unordered groupIds to work which can in turn send groups smaller than the largest.

Author: Since there are no negative Group IDs, {0,0} is not a special case.  All Objects with Location >= Start pass the filter.

present) pass the filter.

The `Largest Object` is defined to be the object with the largest Location
({{location-structure}}) in the track from the perspective of the endpoint

> processing the SUBSCRIBE message.

Reviewer: since the term "LargestObject" is used in multiple places, may be it will benefit from a sub-section so that all the usages refer to the same defintion. It need not be in this PR and we can create a new issue to track, if it helps.

There are 3 types of filters:

Latest Object (0x2): The Start Location is `{Largest Object.Group, Largest Object.Object + 1}`. `Largest Object` is

> communicated in SUBSCRIBE_OK. If no content has been delivered yet, the Start

Reviewer: So that i understand correctly, this means that the current object will never be delivered with filter type `Latest`and the receiver has to wait for the next object to arrive (which may never arrive ?)

I am not sure if we want the above.

Author: I think that can be up to the implementation.  If it's halfway through 5-2 and has the first part of it cached, it can return Largest Object 5-1 in Subscribe OK, and 5-2 is included in Latest.  I was planning to make the explicit in a follow up to this PR if we can get agreement on the general principles here.

Reviewer: One thing i got confused when reading this was, there is a Start (Location) field in the Subscribe message as well. The use of Start Location in this PR refers to publisher's view of filter's start location and not the field in the subscribe message. 

I feel it will help clarify or disambiguate the field vs state in the code on the publisher.

Author: Good Point.  I'll disambiguate this.

> Location is {0, 0}.  There is no End Group - the

Reviewer: ```suggestion
Location is undefined.  There is no End Group - the
```

> Location is the first published or received group.  There is no End Group - the

Author: ```suggestion
Location is {0, 0}.  There is no End Group - the
```

subscription is open ended.  Note that due to network reordering or
prioritization, relays can receive objects with Locations smaller than `Largest
Object` after the SUBSCRIBE is processed, but these objects do not pass the
Latest Object filter.

> AbsoluteStart (0x3):  The Start Location is specified explicitly in the message.

Reviewer: ```suggestion
AbsoluteStart (0x3):  The Start Location is specified explicitly in the Subscribe message.
```

> The Start Location MAY be less than the `Largest Object`. There is no End Group

Reviewer: ```suggestion
The Start Location specified in the Subscribe message MAY be less than the `Largest Object` observed at the publisher. There is no End Group
```

- the subscription is open ended.  To receive any object that is published or
is received after this subscription is processed, a subscriber can use an

> AbsoluteStart filter with Start Location = {0, 0}.

Reviewer: Just to confirm this is a way of supporting all objects ?

Author: Yes, all Objects published or received via subscription pass a filer that starts at {0,0} and has no End Group.

The subscriber specifies a filter on the subscription to allow
the publisher to identify which objects need to be delivered.

All filters have a Start Location and an optional End Group.  Only objects
published or received via a subscription having Locations greater than or

> equal to Start and strictly less than or within the End Group (when

Reviewer 2: Unless there's a reason you're using within, how about equal to?

```suggestion
equal to Start and strictly less than or equal to the End Group (when
```

present) pass the filter.

The `Largest Object` is defined to be the object with the largest Location
({{location-structure}}) in the track from the perspective of the endpoint
processing the SUBSCRIBE message.

There are 3 types of filters:

> Latest Object (0x2): The Start Location is the next object published or received
> via subscription with a Location larger than the `Largest Object`, which is
> communicated in SUBSCRIBE_OK. If no content has been delivered yet, the Start

Author: ```suggestion
Latest Object (0x2): The Start Location is `{Largest Object.Group, Largest Object.Object + 1}`. `Largest Object` is
communicated in SUBSCRIBE_OK. If no content has been delivered yet, the Start
```

> Location is the first published or received group.  There is no End Group - the

Author: > If no content has been delivered yet, the Start Location is the first published or received group.

This is an old rule, but I think this makes it challenging to update this subscription range later, since you might accidentally move it backwards.  Should we change this to say Start Location is {0,0} when Content Exists=0?

> subscription is open ended.  Note that due to network reordering or

Author: Only Latest Object uses Largest Object to create the filter, see the end of the sentence:

> but these objects do not pass the Latest Object filter.

prioritization, relays can receive objects with Locations smaller than `Largest
Object` after the SUBSCRIBE is processed, but these objects do not pass the

> Latest Object filter.

Reviewer 3: Along with the clarified definition of Locations (which is based on comparisons between _integer_ values of groupID and objectID), it means that, in [this example of unordered groupID](https://github.com/moq-wg/moq-transport/issues/892#issuecomment-2763852591), the Relay would implement the third interpretation (forward only the objects whose Location is greater than the "Start Location"). Hence, it would be a filter for the incoming objects received from the upstream subscribe. It requires the Relay to store the Start Location for every subscriber but this is fine.

Dealing with a Subscribe Latest Object (17,0), the Relay who receives from upstream (19,0), (8, 2), (32,0) and (1,1) in this temporal order, will only forward (19,0) and (32,0) downstream.

AbsoluteStart (0x3):  The Start Location is specified explicitly in the message.
The Start Location MAY be less than the `Largest Object`. There is no End Group
- the subscription is open ended.  To receive any object that is published or
is received after this subscription is processed, a subscriber can use an
AbsoluteStart filter with Start Location = {0, 0}.

AbsoluteRange (0x4):  The Start Location and End Group are specified explicitly
in the message. The Start Location MAY be less than the `Largest Object`. If End
Group is the same group specified in Start Location, the remainder of that group
passes the filter.  EndGroup MUST specify the same or a later group than

> specified in `Start`.

Author: You dropped the "MAY be less than Largest Object", I thought that was useful.

published or received via a subscription having Locations greater than or
equal to Start and strictly less than or equal to the End Group (when
present) pass the filter.

The `Largest Object` is defined to be the object with the largest Location
({{location-structure}}) in the track from the perspective of the endpoint
processing the SUBSCRIBE message.

There are 3 types of filters:

Latest Object (0x2): The filter Start Location is `{Largest Object.Group,
Largest Object.Object + 1}` and `Largest Object` is communicated in
SUBSCRIBE_OK. If no content has been delivered yet, the filter Start Location is
{0, 0}. There is no End Group - the subscription is open ended.  Note that due
to network reordering or prioritization, relays can receive Objects with
Locations smaller than  `Largest Object` after the SUBSCRIBE is processed, but
these Objects do not pass the Latest Object filter.

AbsoluteStart (0x3):  The filter Start Location is specified explicitly in the
SUBSCRIBE message. The `Start` specified in the SUBSCRIBE message MAY be less
than the `Largest Object` observed at the publisher. There is no End Group - the

> subscription is open ended.  To receive any Object that is published or is

Author: Yes, every object published or received via subscription will pass a filter with Absolute Start {0,0}, because object.location >= filter.startLocation.

I can change the text to `all Objects that are published or are received...` if that is clearer.

received after this subscription is processed, a subscriber can use an
AbsoluteStart filter with `Start` = {0, 0}.

AbsoluteRange (0x4):  The filer Start Location and End Group are specified
explicitly in the SUBSCRIBE message. The `Start` specified in the SUBSCRIBE
message MAY be less than the `Largest Object` observed at the publisher. If the
specified `EndGroup` is the same group specified in `Start`, the remainder of
that Group passes the filter. `EndGroup` MUST specify the same or a later Group

> than specified in `Start`.

Author: I updated later to larger here and in Subscribe Update.  I think the intent is input validation on the message, so it's clear that larger was the intent.  There's only more more place we use the word _later_ ambiguously, which is in DELIVERY TIMEOUT.  I think we should handle it separately.

Unlike a new subscription, SUBSCRIBE_UPDATE can not cause an Object to be
delivered multiple times.  Like SUBSCRIBE, EndGroup MUST specify the

> same or a larger Object than StartGroup and StartObject.

Reviewer 4: ```suggestion
same or a larger Object id than StartGroup and StartObject.
```

consider clarifying that this is not an object with a larger payload length
